File: solutions/elixir/strain/1/lib/strain.ex
defmodule Strain do
  @doc """
  Given a `list` of items and a function `fun`, return the list of items where
  `fun` returns true.
  """
  @spec keep(list :: list(any), fun :: (any -> boolean)) :: list(any)
  def keep(list, fun), do: kept(list, fun)

  defp kept(list, predicate, kept \\ [])
  defp kept([], _, kept), do: Enum.reverse(kept)
  defp kept([item | list], predicate, kept),
    do: predicate.(item)
      |> then(fn 
        true -> kept(list, predicate, [item | kept])
        _ -> kept(list, predicate, kept)
      end)

  @doc """
  Given a `list` of items and a function `fun`, return the list of items where
  `fun` returns false.
  """
  @spec discard(list :: list(any), fun :: (any -> boolean)) :: list(any)
  def discard(list, fun), do: discarded(list, fun)

  defp discarded(list, predicate, discarded \\ [])
  defp discarded([], _, discarded), do: Enum.reverse(discarded)
  defp discarded([item | list], predicate, discarded),
    do: predicate.(item)
      |> then(fn 
        false -> discarded(list, predicate, [item | discarded])
        _ -> discarded(list, predicate, discarded)
      end)
end

File: solutions/erlang/strain/1/src/strain.erl
-module(strain).

-export([keep/2, discard/2]).

keep(Fn, List) -> keep(Fn, List, []).
keep(Fn, [Item | List], Result) -> 
    case Fn(Item) of
        false -> keep(Fn, List, Result);
        true -> keep(Fn, List, [Item | Result])
    end;
keep(Fn, [], Result) -> lists:reverse(Result).

discard(Fn, List) -> keep(fun(Item) -> not Fn(Item) end, List).
